modules/shared/src/main/scala/org/tessellation/schema/GlobalSnapshotInfo.scala

```diff
@@ -1,20 +1,79 @@
 package org.tessellation.schema
 
+import cats.MonadThrow
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.functor._
+
 import scala.collection.immutable.SortedMap
 
+import org.tessellation.currency.schema.currency.{CurrencyIncrementalSnapshot, CurrencySnapshotInfo}
+import org.tessellation.ext.crypto._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.merkletree.syntax._
+import org.tessellation.merkletree.{MerkleRoot, Proof}
 import org.tessellation.schema.address.Address
 import org.tessellation.schema.balance.Balance
-import org.tessellation.schema.snapshot.SnapshotInfo
+import org.tessellation.schema.snapshot.{SnapshotInfo, StateProof}
 import org.tessellation.schema.transaction.TransactionReference
 import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
 
 import derevo.cats.{eqv, show}
 import derevo.circe.magnolia.{decoder, encoder}
 import derevo.derive
 
 @derive(encoder, decoder, eqv, show)
-case class GlobalSnapshotInfo(
+case class GlobalSnapshotInfoV1(
   lastStateChannelSnapshotHashes: SortedMap[Address, Hash],
   lastTxRefs: SortedMap[Address, TransactionReference],
   balances: SortedMap[Address, Balance]
-) extends SnapshotInfo
+) extends SnapshotInfo[GlobalSnapshotStateProof] {
+  def stateProof[F[_]: MonadThrow: KryoSerializer]: F[GlobalSnapshotStateProof] =
+    GlobalSnapshotInfoV1.toGlobalSnapshotInfo(this).stateProof[F]
+}
+
+object GlobalSnapshotInfoV1 {
+  implicit def toGlobalSnapshotInfo(gsi: GlobalSnapshotInfoV1): GlobalSnapshotInfo =
+    GlobalSnapshotInfo(
+      gsi.lastStateChannelSnapshotHashes,
+      gsi.lastTxRefs,
+      gsi.balances,
+      SortedMap.empty,
+      SortedMap.empty
+    )
+}
+
+@derive(encoder, decoder, eqv, show)
+case class GlobalSnapshotStateProof(
+  lastStateChannelSnapshotHashesProof: Hash,
+  lastTxRefsProof: Hash,
+  balancesProof: Hash,
+  lastCurrencySnapshotsProof: Option[MerkleRoot]
+) extends StateProof
+
+object GlobalSnapshotStateProof {
+  def apply: ((Hash, Hash, Hash, Option[MerkleRoot])) => GlobalSnapshotStateProof = {
+    case (x1, x2, x3, x4) => GlobalSnapshotStateProof.apply(x1, x2, x3, x4)
+  }
+}
+
+@derive(encoder, decoder, eqv, show)
+case class GlobalSnapshotInfo(
+  lastStateChannelSnapshotHashes: SortedMap[Address, Hash],
+  lastTxRefs: SortedMap[Address, TransactionReference],
+  balances: SortedMap[Address, Balance],
+  lastCurrencySnapshots: SortedMap[Address, (Option[Signed[CurrencyIncrementalSnapshot]], CurrencySnapshotInfo)],
+  lastCurrencySnapshotsProofs: SortedMap[Address, Proof]
+) extends SnapshotInfo[GlobalSnapshotStateProof] {
+  def stateProof[F[_]: MonadThrow: KryoSerializer]: F[GlobalSnapshotStateProof] =
+    (
+      lastStateChannelSnapshotHashes.hashF,
+      lastTxRefs.hashF,
+      balances.hashF,
+      lastCurrencySnapshots.merkleTree[F].map(_.map(_.getRoot))
```

Comment:
Is it a second call to `.merkleTree[F]` from the `createArtifact`? If so, please pass here a previously created Merkle tree instead of calculating a new one.

Reply:
`merkleTree` is triggered through `stateProof` method on demand (so it calculate the merkle tree out of data) and it's used for the validation.
the second call is to update the proofs which are part of the `GlobalSnapshotInfo`.
how do you see it combined somehow? I mean, should I pass already calculated merkle tree to the `GlobalSnapshotInfo` instead of calculating it on demand? it would mean as well that we will store calculated merkle tree all the time. I'm not sure if it's correct

Reply:
>I mean, should I pass already calculated merkle tree to the GlobalSnapshotInfo instead of calculating it on demand? it would mean as well that we will store calculated merkle tree all the time. I'm not sure if it's correct

Yes, just pass it from the method that already calculated it.